ktor-core/src/org/jetbrains/ktor/application/Atributes.kt

```diff
@@ -0,0 +1,14 @@
+package org.jetbrains.ktor.application
+
+interface AttributeKey
+
+interface HasAttributes {
+    val attributes: MutableMap<AttributeKey, Any>
+}
+
+suppress("UNCHECKED_CAST")
+inline fun <T: HasAttributes, R> T.getAttributeOrPut(key: AttributeKey, block: () -> R): R = attributes.getOrPut(key, block) as R
```

Comment:
when attributes are backed by concurrenthashmap (often), getOrPut is a wrong thing to do.